File: AMCoreAudio/AMCoreAudioDevice+Formatters.m
//
//  AMCoreAudioDevice+Formatters.m
//  AMCoreAudio
//
//  Created by Ruben Nine on 12/20/12.
//  Copyright (c) 2012 Ruben Nine. All rights reserved.
//

#import "AMCoreAudioDevice+Formatters.h"

@implementation AMCoreAudioDevice (Formatters)

+ (NSString *)formattedSampleRate:(Float64)sampleRate
                   useShortFormat:(BOOL)useShortFormat
{
    NSString *formattedString;

    if (useShortFormat)
    {
        formattedString = [NSString stringWithFormat:@"%.1f kHz", sampleRate * 0.001];
    }
    else
    {
        formattedString = [NSString stringWithFormat:@"%.3f kHz", sampleRate * 0.001];
    }

    return formattedString;
}

+ (NSString *)formattedVolumeInDecibels:(Float32)theVolume
{
    return [NSString stringWithFormat:@"%.1fdB", theVolume];
}

- (NSString *)actualSampleRateFormattedWithShortFormat:(BOOL)useShortFormat
{
    NSString *formattedString = [self.class formattedSampleRate:self.actualSampleRate
                                                 useShortFormat:useShortFormat];

    return formattedString;
}

- (NSString *)numberOfChannelsDescription
{
    UInt32 inputChannels;
    UInt32 outputChannels;

    inputChannels = [self channelsForDirection:kAMCoreAudioDeviceRecordDirection];
    outputChannels = [self channelsForDirection:kAMCoreAudioDevicePlaybackDirection];

    NSString *formattedString = [NSString stringWithFormat:NSLocalizedString(@"%d in/ %d out", nil),
                                 inputChannels,
                                 outputChannels];

    return formattedString;
}

- (NSString *)latencyDescription
{
    NSMutableString *formattedString;
    UInt32 inLatency;
    UInt32 outLatency;
    UInt32 inSafetyOffsetLatency;
    UInt32 outSafetyOffsetLatency;
    UInt32 totalInLatency;
    UInt32 totalOutLatency;

    formattedString = [NSMutableString string];

    inLatency = [self deviceLatencyFramesForDirection:kAMCoreAudioDeviceRecordDirection];
    outLatency = [self deviceLatencyFramesForDirection:kAMCoreAudioDevicePlaybackDirection];
    inSafetyOffsetLatency = [self deviceSafetyOffsetFramesForDirection:kAMCoreAudioDeviceRecordDirection];
    outSafetyOffsetLatency = [self deviceSafetyOffsetFramesForDirection:kAMCoreAudioDevicePlaybackDirection];

    totalInLatency = inLatency + inSafetyOffsetLatency;
    totalOutLatency = outLatency + outSafetyOffsetLatency;

    if (totalInLatency > 0)
    {
        [formattedString appendFormat:NSLocalizedString(@"%.1fms in", nil),
         (double)totalInLatency / self.nominalSampleRate * 1000];
    }

    if (totalOutLatency > 0)
    {
        if (formattedString.length > 0)
        {
            [formattedString appendString:NSLocalizedString(@"/ ", nil)];
        }

        [formattedString appendFormat:NSLocalizedString(@"%.1fms out", nil),
         (double)totalOutLatency / self.nominalSampleRate * 1000];
    }

    return [formattedString copy];
}

@end

File: AMCoreAudio/AMCoreAudioDevice+Formatters.swift
//
//  AMCoreAudioDevice+Formatters.swift
//  AMCoreAudio
//
//  Created by Ruben Nine on 18/01/16.
//  Copyright © 2016 9Labs. All rights reserved.
//

import Foundation

@available(*, deprecated, message="this function will be removed in version 3.0") extension AMCoreAudioDevice {

    public final class func formattedSampleRate(sampleRate: Float64, useShortFormat: Bool) -> String {
        if (useShortFormat) {
            return String(format: NSLocalizedString("%.1f kHz", comment: ""),
                          sampleRate * 0.001)
        } else {
            return String(format: NSLocalizedString("%.3f kHz", comment: ""),
                          sampleRate * 0.001)
        }
    }

    public final class func formattedVolumeInDecibels(volume: Float32) -> String {
        return String(format: NSLocalizedString("%.1fdB", comment: ""), volume)
    }

    public final func actualSampleRateFormattedWithShortFormat(useShortFormat: Bool) -> String {
        return self.dynamicType.formattedSampleRate(actualSampleRate() ?? 0, useShortFormat: useShortFormat)
    }

    public final func numberOfChannelsDescription() -> String {
        let inputChannels = channelsForDirection(.Recording) ?? 0
        let outputChannels = channelsForDirection(.Playback) ?? 0

        return String(format: NSLocalizedString("%d in/ %d out", comment: ""),
                      inputChannels,
                      outputChannels
        )
    }

    public final func latencyDescription() -> String {
        let inLatency = deviceLatencyFramesForDirection(.Recording) ?? 0
        let outLatency = deviceLatencyFramesForDirection(.Playback) ?? 0
        let inSafetyOffsetLatency = deviceSafetyOffsetFramesForDirection(.Recording) ?? 0
        let outSafetyOffsetLatency = deviceSafetyOffsetFramesForDirection(.Playback) ?? 0

        let totalInLatency = inLatency + inSafetyOffsetLatency
        let totalOutLatency = outLatency + outSafetyOffsetLatency

        var formattedString = ""

        if totalInLatency > 0 {
            formattedString.appendContentsOf(String(format: NSLocalizedString("%.1fms in", comment: ""),
                Float64(totalInLatency) / (nominalSampleRate() ?? 0) * 1000)
            )
        }

        if totalOutLatency > 0 {
            if !formattedString.isEmpty {
                formattedString.appendContentsOf(NSLocalizedString("/ ", comment: ""))
            }

            formattedString.appendContentsOf(String(format: NSLocalizedString("%.1fms out", comment: ""),
                Float64(totalOutLatency) / (nominalSampleRate() ?? 0) * 1000)
            )
        }

        return formattedString
    }
}
